Make Author block selector to display all users instead of just 10

wojtekn opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

In this PR, I propose to allow for displaying an unlimited number of users in Author block settings.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

It fixes issue that was reported: https://github.com/WordPress/gutenberg/issues/44688

Currently, if a site has more than ten authors, user who edits a post, can't see all of them when he edits an Author block. With that change, they can choose any author user who exists on the site.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

It uses `-1` value option for `per_page` prop, so `fetchAllMiddleware` will help with getting the whole list of author users. While adding that change, I extract the query params to constant to match the code style of other blocks.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Ensure that site has more than 10 author users
2. Create a post.
3. Insert an author block.
4. Check the Block settings.
5. Confirm that all authors are available in the dropdown.

## Screenshots or screencast <!-- if applicable -->

<img width="326" alt="Screen Shot 2022-11-09 at 10 50 53" src="https://user-images.githubusercontent.com/727413/200903796-431054cd-0624-43f2-855c-f102c29d8cda.png">

wojtekn committed:
Make Author block selector to display all users instead of just 10

ramonjd reviewed (approved):
LGTM. Thank you!


All users with author privileges are in my list, and I can select them.

<img width="326" alt="Screen Shot 2022-11-10 at 5 10 46 am" src="https://user-images.githubusercontent.com/6458278/200907673-4c62438e-1094-4cdf-942e-bf24bac99b68.png">

Looks good on frontend.

<img width="440" alt="Screen Shot 2022-11-10 at 5 10 54 am" src="https://user-images.githubusercontent.com/6458278/200907668-eb5eac5e-4027-4e15-97a2-cf120e75716b.png">

noisysocks commented:
I am wondering if we should set the `per_page` to `50` to match the Post Author dropdown in the editor sidebar.

https://github.com/WordPress/gutenberg/blob/fab909c13ec2f98fa160dd8cdc9f4d95dd6066c0/packages/editor/src/components/post-author/constants.js#L1-L6

I'm not sure why it's `50` though to be honest. It used to be `100` before https://github.com/WordPress/gutenberg/pull/33695.

ramonjd commented:
> I am wondering if we should set the per_page to 50 to match the Post Author dropdown in the editor sidebar.

Good point. I guess there are sites out there with 100s of authors, which might not lead to a great UX if they are all added to the dropdown.

But what if we wanted to add an author to an Author block that doesn't end up on the list due to the limit? 🤔

wojtekn commented:
@noisysocks I've searched for usages of `getUsers(` across the other blocks, and I saw in other places we follow the `-1` pattern.

I'm not sure what the reason was to use explicit limit of 50, one thought is that at that time we didn't have that `fetchAllMiddleware`, or author was not aware of it and used 50 as a high number that would fix 99% cases?

wojtekn commented:
@noisysocks @ramonjd the comment from the linked issue provides more insight: https://github.com/WordPress/gutenberg/pull/33695#discussion_r677249877 - we use a combo box component when there are more than 25 authors, so it we should be good with UX experience here.

noisysocks commented:
So it looks like the `PostAuthor` component in the `@wordpress/editor` package handles all of this perfectly which is that it uses a `SelectControl` if there are fewer than 25 users and a `ComboboxControl` if there are more than 25 users. When using a `ComboboxControl` it will perform new REST API fetches as the user types into the text input.

https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/post-author/index.js

I think the ideal solution is to move this `PostAuthor` component into the `@wordpress/block-editor` package (you'll need to ensure backwards compatibility) and then change all of the blocks that have a user dropdown to use this component.

A more temporary fix is to bump the `per_page` query up from `10` to `100`.

wojtekn committed:
Change from unlimited to 100

I am changing the limit from -1 to 100 as for now, to fix issue
for most cases, and at the same time, to avoid introducing
performance problems.

Additionally, I include support for Combobox control for cases
with more than 25 users to keep it aligned with other parts of the
interface.

wojtekn commented:
@noisysocks I've changed `-1` to `100` for now and introduced a ComboBox control to align the UX with other places.

ramonjd reviewed (approved):
Smoked tested after `100` limit. 

Thanks for updating! 🚀

bangank36 reviewed (commented):
LGTM, I liked the way you clean up the `onChange`callback

noisysocks commented:
Thanks @wojtekn! This looks like a good interim fix for me until we can refactor `PostAuthor` per my comment above.

mtias commented:
Were there any performance checks done against this? This endpoint used to suffer a bit and made sites with many authors take a long time to load.

ramonjd commented:
I'm inspecting the endpoint `/index.php?rest_route=%2Fwp%2Fv2%2Fusers&context=edit&who=authors&per_page=100&_locale=user` using 1, 50 and 200 users (the limit introduced in this PR is 100) on fast and slow connection emulation.

I see little or no visible difference when loading the editor, and the total endpoint load time  – from queued to content download – don't appear to differ for all author counts.

I added 10 Post author blocks to a page (there's only one request), and the following requests are 200 and 1 user(s) in order. 

<img width="1281" alt="Screen Shot 2022-11-24 at 3 09 37 pm" src="https://user-images.githubusercontent.com/6458278/203692295-5c5a7c3f-0dd7-4879-b32c-65a2d7945559.png">

For every page load the users endpoint is called at least 3 times:

1. For the logged-in user
2. For the Post panel author drop down
3. For the Post author block settings drop down

On average I'm seeing a combined performance hit of 10-20 milliseconds between the minimum number of authors (1) and the maximum (50 for the Post author drop down, and 100 for the Post author block drop down).

#### 1 author
<img width="1307" alt="Screen Shot 2022-11-24 at 3 13 12 pm" src="https://user-images.githubusercontent.com/6458278/203693532-7b2718e6-d075-401f-bcbc-493477394bab.png">

#### 200 authors
<img width="1308" alt="Screen Shot 2022-11-24 at 3 15 12 pm" src="https://user-images.githubusercontent.com/6458278/203693541-920dff53-ee29-4c44-8814-e552702711f5.png">

All of this of course depends on server and network capacity.

By the way here's what the `ComboboxControl` looks like when displaying the maximum of 100 users:

![2022-11-24 15 05 50](https://user-images.githubusercontent.com/6458278/203691836-2c2650f4-64b9-41fc-baba-1cddca25931a.gif)

Do folks have a more reliable way of measuring this?
